Reference concentration for surface species when calculating Kc

davidfarinajr opened the pull request:
<!--
Thanks for contributing a pull request! Please try to provide as much detail as possible to help the reviewer understand your work.
You can also add the appropriate labels to describe the topic of the pull request and the type of changes you're making.
-->

### Motivation or Problem
This is related to issue https://github.com/ReactionMechanismGenerator/RMG-Py/issues/1937
We currently assume ideal gas mixture to determine Kc.  We should be using different reference concentrations for gas and surface species.  

### Description of Changes
Use reference concentration of (surface site density) for surface site species and ideal gas (P/RT) for gas species.  That way, gas concentration scales with temperature and surface species concentration is independent of temperature.  Therefore, we use the change in mols of gas in the reaction to determine Kc (instead of len(reactants) - len(products)).  However, the species object needs a molecule to determine its phase (if it contains surface site), so I added a try and except for this.

### Testing
I added unit test for determining Kc for a surface reaction. I also modified the gas Kc test to test reactions with species that contain and don't contain a molecule.

### Reviewer Tips
I don't really like my try and except block.  I'm interested in suggestions to improve this.

<!--
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with master?
-->

skrsna commented on rmgpy/reaction.py:
```diff
@@ -560,22 +560,37 @@ def get_equilibrium_constant(self, T, type='Kc'):
+        ``Kc`` for concentrations (default), or ``Kp`` for pressures.  This
+        function assumes a reference pressure of 1e5 Pa for gas phases species
+        and uses the ideal gas law to determine reference concentrations. For 
+        solid surface site species, a reference concentration of 1 is assumed. 
```
added the reference from Cantera's source code https://github.com/Cantera/cantera/blob/master/include/cantera/kinetics/InterfaceKinetics.h#L467

davidfarinajr commented on rmgpy/reaction.py:
```diff
@@ -560,22 +560,37 @@ def get_equilibrium_constant(self, T, type='Kc'):
+        # of gas will be used to determine Kc and Kp since solid species have reference
+        # concentration of 1
+        try:
+            gas_reacts = [spcs for spcs in self.reactants if not spcs.is_surface_site()]
```
Just realized I need to change `is_surface_site()` to `contains_surface_site()`

nateharms commented on rmgpy/reaction.py:
```diff
@@ -560,22 +560,37 @@ def get_equilibrium_constant(self, T, type='Kc'):
+        try:
+            gas_reacts = [spcs for spcs in self.reactants if not spcs.contains_surface_site()]
+            gas_prods = [spcs for spcs in self.products if not spcs.contains_surface_site()]
+        except IndexError:
```
Okay, after looking through `species.py`, it looks like the `IndexError` is appropriate here. I understand why you'd feel it was wrong at first, I had to do some digging myself.

rwest commented:
This looks like it was approved three times three months ago. @davidfarinajr please could you rebase it once more and post in slack to try to get it merged? thanks

davidfarinajr commented:
I'm still not confident about using 1 for reference concentration of surface species.  1 is typically used for solids and liquids since their concentration does't change in the reaction, but the concentration of our adsorbates changes a lot (ranges from 0 to the surface site density).   I don't think we should be using ideal gas to determine reference concentration (as our surface concentrations don't change with temp), but perhaps we should be using the surface site density as the reference instead of 1?

rwest commented:
Ok. Maybe we consult @cfgoldsmith ?

rwest commented:
Could also compare with Cantera.

skrsna commented:
AFAIK, David's code is based on Cantera's code so I don't know how I feel about comparing it to Cantera's values.

davidfarinajr commented:
According to this review paper, it looks like this topic, of what to use for standard concentrations of adsorbates for multi dentate adsorptions reactions (where the surface species don't cancel out), is somewhat disputed:
https://pubs.acs.org/doi/pdf/10.1021/es305180e

davidfarinajr commented:
After discussion with @cfgoldsmith , we decided to use `surface site density` for reference concentration instead of 1.  I have rebased and made this change.  I added `surface_site_density` as a parameter to the `get_equilibrium_constant` method with a default value of `2.5e-05` mol/m^2 (approximately the value for Pt111).

[rmgcat_reference_surface_concentration.pdf](https://github.com/ReactionMechanismGenerator/RMG-Py/files/5138752/rmgcat_reference_surface_concentration.pdf)

rwest commented on rmgpy/reaction.py:
```diff
@@ -555,27 +555,50 @@ def get_free_energy_of_reaction(self, T):
+        ``Kc`` for concentrations (default), or ``Kp`` for pressures.  This
+        function assumes a reference pressure of 1e5 Pa for gas phases species
+        and uses the ideal gas law to determine reference concentrations. For 
+        surface site species, a reference concentration of 1 is assumed. 
```
Docstring needs updating to reflect latest changes.

rwest commented on rmgpy/reaction.py:
```diff
@@ -555,27 +555,50 @@ def get_free_energy_of_reaction(self, T):
         P0 = 1e5
+        # Determine the number of gas phase reactants and products. The change in mols
+        # of gas will be used to determine Kc and Kp since solid species have reference
+        # concentration of 1
```
update comment

rwest commented on rmgpy/reaction.py:
```diff
@@ -555,27 +555,50 @@ def get_free_energy_of_reaction(self, T):
+            gas_reacts = len([spcs for spcs in self.reactants if not spcs.contains_surface_site()])
+            gas_prods = len([spcs for spcs in self.products if not spcs.contains_surface_site()])
+        except IndexError:
+            logging.warning(f"Species do not have an rmgpy.molecule.Molecule "  
```
```suggestion
            logging.warning("Species do not have an rmgpy.molecule.Molecule "  
```

rwest commented on rmgpy/reaction.py:
```diff
@@ -555,27 +555,50 @@ def get_free_energy_of_reaction(self, T):
+
+        surf_reacts = len(self.reactants) - gas_reacts
+        surf_prods = len(self.products) - gas_prods
+        v_surf = surf_prods - surf_reacts # change in mols of surface spcs
```
not a big fan of these variable names. I guess you were going for "nu" but the v makes me start asking "why vee? is it volume?". Maybe `dN_gas` or `delta_stoichiometry_gas` or `change_in_moles_of_gas`?

rwest commented on rmgpy/reaction.py:
```diff
@@ -555,27 +555,50 @@ def get_free_energy_of_reaction(self, T):
+            logging.warning(f"Species do not have an rmgpy.molecule.Molecule "  
+                            "Cannot determine phases of species. We will assume "
+                            "ideal gas mixture when calculating Kc and Kp.")
+            gas_reacts = len(self.reactants)
```
I'd rather it be called `gas_reactants` or even `number_of_gas_reactants`  (same for `gas_prods` below)

rwest commented on rmgpy/reactionTest.py:
```diff
@@ -257,6 +257,18 @@ def test_methyl_adsorption_sticking_coefficient(self):
+        Test the Reaction.get_equilibrium_constant() method.
+        """
+        Tlist = numpy.arange(400.0, 1600.0, 200.0, numpy.float64)
+        Kclist0 = [float(v) for v in
```
why do this float(v) for v in list of strings, instead of just doing a list of floats?

davidfarinajr replied:
Not sure, I notice other tests were written this way, so I followed that

davidfarinajr commented:
Thanks for your review @rwest ! I believe I addressed all of your comments, could you take another look?

mazeau reviewed (dismissed):
This looks fine to me!

rwest reviewed (dismissed):
I just reviewed and made a few tiny changes while rebasing to the latest head.
If the tests pass I think this should be merged.

davidfarinajr commented:
Thanks Richard!

rwest reviewed (approved):
Regarding the discussion earlier in the comment thread, @davidfarinajr,  @cfgoldsmith, and I agreed we should be using surface site density as the appropriate reference. I have checked the implementation, and reviewed the code. All tests pass.
